src/qibolab/instruments/qm/config/config.py

```diff
@@ -0,0 +1,125 @@
+from dataclasses import dataclass, field
+from typing import Union
+
+from qibolab.components.configs import IqConfig, OscillatorConfig
+from qibolab.pulses import Pulse
+
+from ..components import OpxOutputConfig, QmAcquisitionConfig, QmChannel
+from .devices import *
+from .elements import *
+from .pulses import *
+
+__all__ = ["QmConfig"]
+
+
+@dataclass
+class QmConfig:
+    """Configuration for communicating with the ``QuantumMachinesManager``.
+
+    Contains nested ``dataclass`` objects and is serialized using ``asdict``
+    to be sent to the instrument.
+    """
+
+    version: int = 1
+    controllers: dict[str, Controller] = field(default_factory=dict)
+    octaves: dict[str, Octave] = field(default_factory=dict)
+    elements: dict[str, Element] = field(default_factory=dict)
+    pulses: dict[str, Union[QmPulse, QmAcquisition]] = field(default_factory=dict)
+    waveforms: dict[str, Waveform] = field(default_factory=dict)
+    digital_waveforms: dict = field(
+        default_factory=lambda: {"ON": {"samples": [(1, 0)]}}
+    )
+    integration_weights: dict = field(default_factory=dict)
+    mixers: dict = field(default_factory=dict)
+
+    def add_controller(self, device: str):
+        if device not in self.controllers:
+            self.controllers[device] = Controller()
+
+    def add_octave(self, device: str, connectivity: str):
+        if device not in self.octaves:
+            self.add_controller(connectivity)
+            self.octaves[device] = Octave(connectivity)
+
+    def configure_dc_line(self, channel: QmChannel, config: OpxOutputConfig):
+        controller = self.controllers[channel.device]
+        controller.analog_outputs[str(channel.port)] = config
+        self.elements[channel.logical_channel.name] = DcElement(channel.serial)
+
+    def configure_iq_line(
+        self, channel: QmChannel, config: IqConfig, lo_config: OscillatorConfig
+    ):
+        port = channel.port
+        octave = self.octaves[channel.device]
+        octave.RF_outputs[str(port)] = OctaveOutput.from_config(lo_config)
+        self.controllers[octave.connectivity].add_octave_output(port)
```

Comment:
> Here I'm just puzzled, but I should think about it more carefully (so, the comment is also a reminder to myself).
> 
> It's true that you're configuring an IQ line, and that's a modular operation. However, in order to configure the line, you need the octave to be already partially registered. But it's configuration as an LO is coming only here.
> 
> I'd have expected that either the octave had to be fully registered in advance, or it would be fully configured here (which doesn't seem to be possible, since the connectivity might involve multiple lines).

The Octave is registered in advance calling `config.add_octave`. The output ports are registered by this function which I believe has all the required information (LO configuration, IF, etc.).

> Also, what it happens if the octave is configured on multiple lines, with incompatible OscillatorConfigs?
I guess not all the lines are distinguished by (dev, port), since that should be different only for physically different lines. But maybe I'm just wrong...

Here you are partly right. The idea is that the user is aware of how many LOs are available in the instrument and creates the components appropriately to avoid collisions: https://github.com/qiboteam/qibolab_platforms_qrc/blob/0.2-qw11qD/qw11qD/parameters.json. I guess if this is not done properly, they may end up overwriting some LO configuration by mistake. For Octaves in particular, this is a bit trickier because as discussed before ports 2, 3 and 4, 5 share LOs even though we are allowed to pass different LO frequencies in the config (one of them is ignored). In principle, this is a problem from the qm-qua side and I hope they will raise appropriate errors in later releases (that's not the case so far). Otherwise the only other solution is to raise the errors ourselves.

Reply:
Ok, I see it's not trivial at all, and there is no urgent need to change the current status.

Thanks for the explanation, I'll just keep that in mind!